types/math/dec_test.go

```diff
@@ -89,28 +137,437 @@ func TestDec(t *testing.T) {
+
+	c, err := a.Add(b)
+	require.NoError(t, err)
+
+	d, err := c.Sub(b)
+	require.NoError(t, err)
+
+	require.True(t, a.IsEqual(d))
+}
+
+// TODO: Revive this property test with appropriate errors once we need Mul
+// Property: a * (b + c) == (a * b) + (a * c)
+// func testAddMulDistributes(t *rapid.T) {
+// 	a := genDec.Draw(t, "a").(Dec)
+// 	b := genDec.Draw(t, "b").(Dec)
+// 	c := genDec.Draw(t, "c").(Dec)
+
+// 	// a * (b + c)
+// 	d, err := b.Add(c)
+// 	require.NoError(t, err)
+
+// 	e, err := a.Mul(d)
+// 	require.NoError(t, err)
+
+// 	// (a * b) + (a * c)
+// 	f, err := a.Mul(b)
+// 	require.NoError(t, err)
+
+// 	g, err := a.Mul(c)
+// 	require.NoError(t, err)
+
+// 	h, err := f.Add(g)
+// 	require.NoError(t, err)
+
+// 	// Error value
+// 	// TODO: Hmm it seems we can get arbitrarily large errors here...what to do?
+// 	errDec, err := h.Sub(e)
+// 	require.NoError(t, err)
+
+// 	t.Log(errDec.String())
+
+// 	maxErr, err := NewDecFromString("1000")
+// 	require.NoError(t, err)
+
+// 	require.GreaterOrEqual(t, 0, errDec.Cmp(maxErr))
+// }
+
+// TODO: Revive this property test with appropriate errors once we need Mul and
+// Quo
+// Property: ((a div b) * b) + Rem(a, b) == a
+// func testMulQuoIntegerRem(t *rapid.T) {
+// 	decNotZero := func(d Dec) bool { return !d.IsZero() }
+// 	a := genDec.Draw(t, "a").(Dec)
+// 	b := genDec.Filter(decNotZero).Draw(t, "b").(Dec)
+// 	t.Logf("a: %v", a)
+// 	t.Logf("b: %v", b)
+
+// 	c, err := a.QuoInteger(b)
+// 	if err != nil {
+// 		// Catch the case where the precision means this is impossible
+// 		// TODO: Make sure this is happening because of precision issues
+// 		require.Contains(t, err.Error(), "division impossible")
+// 		t.SkipNow()
+// 	}
+// 	require.NoError(t, err)
+// 	t.Logf("a div b: %v", c)
+
+// 	d, err := c.Mul(b)
+// 	require.NoError(t, err)
+// 	t.Logf("(a div b) * b: %v", d)
+
+// 	e, err := a.Rem(b)
+// 	require.NoError(t, err)
+// 	t.Logf("a.Rem(b): %v", e)
+
+// 	f, err := d.Add(e)
+// 	require.NoError(t, err)
+// 	t.Logf("f: %v", f)
+
+// 	// require.True(t, a.IsEqual(f))
+// 	errDec, err := f.Sub(a)
+// 	require.NoError(t, err)
+// 	t.Logf("errDec: %v", errDec)
+
+// 	maxErr, err := NewDecFromString("100")
+// 	require.NoError(t, err)
+
+// 	// errDec <= maxErr
+// 	require.GreaterOrEqual(t, 0, errDec.Cmp(maxErr))
+// }
+
+// Property: a * 0 = 0
+func testMulZero(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	zero := Dec{}
+
+	c, err := a.Mul(zero)
+	require.NoError(t, err)
+	require.True(t, c.IsZero())
+}
+
+// Property: a/a = 1
+func testSelfQuo(t *rapid.T) {
+	decNotZero := func(d Dec) bool { return !d.IsZero() }
+	a := genDec.Filter(decNotZero).Draw(t, "a").(Dec)
+	one := NewDecFromInt64(1)
+
+	b, err := a.Quo(a)
+	require.NoError(t, err)
+	require.True(t, one.IsEqual(b))
+}
+
+// Property: a/1 = a
+func testQuoByOne(t *rapid.T) {
+	decNotZero := func(d Dec) bool { return !d.IsZero() }
+	a := genDec.Filter(decNotZero).Draw(t, "a").(Dec)
```

Comment:
we don't need the filter for this test

Reply:
ah right, got mixed up between 0/x and x/0